src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBitsRoslyn.cs

```diff
@@ -0,0 +1,499 @@
+            {
+                fractionalShift = fractionalDenominatorBits - fractionalNumeratorBits;
+            }
+
+            if (fractionalShift > 0)
+            {
+                fractionalNumerator.ShiftLeft(fractionalShift);
+            }
+
+            uint requiredFractionalBitsOfPrecision = requiredBitsOfPrecision - integerBitsOfPrecision;
+            uint remainingBitsOfPrecisionRequired = requiredFractionalBitsOfPrecision;
+
+            if (integerBitsOfPrecision > 0)
+            {
+                // If the fractional part of the mantissa provides no bits of precision  
+                // and cannot affect rounding, we can just take whatever bits we got from  
+                // the integer part of the mantissa.  This is the case for numbers like  
+                // 5.0000000000000000000001, where the significant digits of the fractional  
+                // part start so far to the right that they do not affect the floating  
+                // point representation.  
+                //  
+                // If the fractional shift is exactly equal to the number of bits of  
+                // precision that we require, then no fractional bits will be part of the  
+                // result, but the result may affect rounding.  This is e.g. the case for  
+                // large, odd integers with a fractional part greater than or equal to .5.  
+                // Thus, we need to do the division to correctly round the result.  
+                if (fractionalShift > remainingBitsOfPrecisionRequired)
+                {
+                    return ConvertBigIntegerToFloatingPointBits(
+                        ref integerValue,
+                        in info,
+                        integerBitsOfPrecision,
+                        fractionalDigitsPresent != 0
+                    );
+                }
+
+                remainingBitsOfPrecisionRequired -= fractionalShift;
+            }
+
+            // If there was no integer part of the mantissa, we will need to compute the  
+            // exponent from the fractional part.  The fractional exponent is the power  
+            // of two by which we must multiply the fractional part to move it into the  
+            // range [1.0, 2.0).  This will either be the same as the shift we computed  
+            // earlier, or one greater than that shift:  
+            uint fractionalExponent = fractionalShift;
+
+            if (BigInteger.Compare(ref fractionalNumerator, ref fractionalDenominator) < 0)
+            {
+                fractionalExponent++;
+            }
+
+            fractionalNumerator.ShiftLeft(remainingBitsOfPrecisionRequired);
+
+            BigInteger.DivRem(ref fractionalNumerator, ref fractionalDenominator, out BigInteger bigFractionalMantissa, out BigInteger fractionalRemainder);
+            ulong fractionalMantissa = bigFractionalMantissa.ToUInt64();
+            bool hasZeroTail = fractionalRemainder.IsZero();
```

Comment:
This, and other cases of `hasZeroTail` aren't quite correct yet. `NumberBuffer` needs to carry an additional parameter that tracks whether any digits that were parsed, but not stored in the buffer (so digits past 113 for single and past 768 for double), were non-zero to ensure that rounding is always correct/accurate.

Reply:
Resolved.